### DISCUSSION TITLE
Moe Memos 2 Roadmap

### DISCUSSION THREAD

**Discussion by mudkipme** [mudkipme/MoeMemos#128] [Ideas]
Moe Memos 2 is set to be a major new release and including many long awaited improvements.

Here are some of the new features I've thought of so far:

* Platform support: Android 8+, iOS 17+, macOS Sonoma (Catalyst), and visionOS.
	* Due to the usage of new SwiftUI APIs and SwiftData, support for iOS 15 and 16 will be dropped. However, I'll maintain minor releases for v1.x on TestFlight in case new server versions break compatibility.
	* The layout will be optimized for Android tablets.
	* The macOS app will use the Mac idiom, so it can feel more native than the current iOS app on Mac.
* Memos server support: Only the Access Token will be used internally. Open ID/Cookies will no longer be in use. While Username/Password login will still be available, it is not recommended due to the default expiration time.
	- Thus, only Memos 0.15 or later versions will be compatible.
* Multi-user/instance capabilities:
	* Users can add multiple accounts (from the same or different Memos servers) and switch among them.
	* If an Access Token expires or is deleted, users can log back into the same account, provided the host/account ID remains unchanged.
* Offline/Local support:
	* Users can create a "local account" to use Moe Memos independently of a Memos server. (Support is limited to one local account)
	* Logged-in users can also access Moe Memos offline. Memos that are created, edited, or deleted offline will be stored locally and synchronized with the server once online.
	* In cases of conflict (e.g., a memo edited or deleted both locally and on the server), a new memo will be created.
	* A local account can be merged into a Memos account, but the reverse is not possible. After merging, all local memos will appear on the Memos server while retaining their original creation dates.
	* If a memo fails to synchronize (for instance, an attachment is too large for upload), it will be labeled as "unsynced."
* Attachments:
	* Users can save attachments beyond just images within a memo and can also save files from other apps.
	* There's an option to use either the original or compressed image format.
* Memo editor:
	* Features a Markdown toolbar with headlines, font styles, links, etc.
	* Syntax highlighting on iOS, powered by [Runestone](https://github.com/simonbs/Runestone), is under consideration.
* Widgets:
	* Memo Graph / Memory Widget will be added to Android version.
	* Lock screen widgets will be added to iPhone version.
* Security:
	* Users will have the option to secure Moe Memos using biometric verification.
	* Plain HTTP server support on iOS is being considered, depending on Apple's approval.
* Reading memos:
	* Users can decide whether or not to collapse lengthy memos.
	* There will be a dedicated memo page, and tapping the Memory widget will direct users to that memo.
	* Users can share memo links or view them on the Memos web page.
* Tags:
	* Support for multi-level tags will be introduced.
* Miscellaneous
	* Several optional icons will be available on iOS.
	* A "buy me a coffee"-style In-App Purchase option will be added to iOS. However, all features will remain completely free and open source.

I've already begun work on this new version, though I'm not sure when I can release it. I'm aiming for early next year, but there's a strong chance I might not meet that target.

**Comment by PylotLight** [comment]
How'd you go with this as it sounds amazing!?

**Comment by denastria** [comment]
很期待

**Comment by denastria** [comment]
很期待

**Comment by DrSpaldo** [comment]
@mudkipme I was wondering if use through [Cloudflare Zero Trust](https://developers.cloudflare.com/cloudflare-one/identity/service-tokens/) could be included in the upcoming version? Some people in other similar apps call it extra headers. 

There is a bit of a discussion here for the Paperless NGX app; https://github.com/paulgessinger/swift-paperless/issues/11

The source is available for that app, so maybe you could have a look how it was done? Hopefully @paulgessinger wont mind...

**Comment by ocean-haiyang** [comment]
When can we expect offline support?

**Comment by WACOMalt** [comment]
Please add a widget for showing/interacting with a single memo. I have a daily todo list that I use in keep. I can't do this with Moe memos. It's the only thing keeping me using Google keep.

**Comment by DrSpaldo** [comment]
Any thoughts on a settings / options area? Maybe include different icons, light/dark settings, etc
